pkg/kubelet/cm/dra/plugin/plugin_manager.go

```diff
@@ -0,0 +1,503 @@
+	pm.cancel(errors.New("Stop was called"))
+
+	// Close all connections, otherwise gRPC keeps doing things in the background.
+	// Also cancel all pending wiping.
+	for driverName, plugins := range pm.store {
+		workerArg := timedworkers.NewWorkArgs(driverName, "")
+		pm.pendingWipes.CancelWork(logger, workerArg.KeyFromWorkArgs())
+		for _, plugin := range plugins {
+			if err := plugin.conn.Close(); err != nil {
+				logger.Error(err, "Closing gRPC connection", "driverName", plugin.driverName, "endpoint", plugin.endpoint)
+			}
+		}
+	}
+}
+
+// wipeResourceSlices deletes ResourceSlices of the node, optionally just for a specific driver.
+//
+// It gets called in a stand-alone goroutine at kubelet startup and as callback
+// of a TimedWorkersQueue. In both cases the caller has no way of handling errors,
+// so wipeResourceSlices must implement it's own retry mechanism.
+//
+// Can be canceled by canceling the context.
+func (pm *PluginManager) wipeResourceSlices(ctx context.Context, driver string) {
+	if pm.kubeClient == nil {
+		return
+	}
+	logger := klog.FromContext(ctx)
+
+	backoff := wait.Backoff{
+		Duration: time.Second,
+		Factor:   2,
+		Jitter:   0.2,
+		Cap:      5 * time.Minute,
+		Steps:    100,
+	}
+
+	// Error logging is done inside the loop. Context cancellation doesn't get logged.
+	_ = wait.ExponentialBackoffWithContext(ctx, backoff, func(ctx context.Context) (bool, error) {
+		node, err := pm.getNode()
+		if apierrors.IsNotFound(err) {
+			return false, nil
+		}
+		if err != nil {
+			logger.Error(err, "Unexpected error checking for node")
+			return false, nil
+		}
+		fieldSelector := fields.Set{resourceapi.ResourceSliceSelectorNodeName: node.Name}
+		if driver != "" {
+			fieldSelector[resourceapi.ResourceSliceSelectorDriver] = driver
+		}
+
+		err = pm.kubeClient.ResourceV1beta1().ResourceSlices().DeleteCollection(ctx, metav1.DeleteOptions{}, metav1.ListOptions{FieldSelector: fieldSelector.String()})
+		switch {
+		case err == nil:
+			logger.V(3).Info("Deleted ResourceSlices", "fieldSelector", fieldSelector)
+			return true, nil
+		case apierrors.IsUnauthorized(err):
+			// This can happen while kubelet is still figuring out
+			// its credentials.
+			logger.V(5).Info("Deleting ResourceSlice failed, retrying", "fieldSelector", fieldSelector, "err", err)
+			return false, nil
+		default:
+			// Log and retry for other errors.
+			logger.V(3).Info("Deleting ResourceSlice failed, retrying", "fieldSelector", fieldSelector, "err", err)
+			return false, nil
+		}
+	})
+}
+
+// GetDRAPlugin returns a wrapper around those gRPC methods of a DRA
+// driver kubelet plugin which need to be called by kubelet. The wrapper
+// handles gRPC connection management and logging. Connections are reused
+// across different calls.
+//
+// It returns an informative error message including the driver name
+// with an explanation why the driver is not usable.
+func (pm *PluginManager) GetDRAPlugin(driverName string) (*Plugin, error) {
```

Comment:
Does it make sense to rename this type to `DRAPluginManager`, the returned type to `DRAPlugin` and then have this function just be `called GetPlugin()`?

Reply:
Looks good to me. I'll update the renaming commit...